### DISCUSSION TITLE
Selecting grains by mouse

### DISCUSSION THREAD

**Discussion by rtaufner** [mtex-toolbox/mtex#831] [Ask Anything] [answered]
Hello Mtexers, hopefully everyone is safe and sound!

I am trying to manually select some grains around a diopside porphyroclast to check any host-controlled orientation. The code I am using is: 
`%% Select specific grains (more than one grain is allowed)`

`plot (grains)`
`selectInteractive(grains,'lineColor','gold')`

`% this simulates a mouse click`
`pause(0.1)`
`simulateClick(9000,3500)`
`pause(0.1)`

`global indSelected;`
`grains(indSelected)`

`hold on`
`plot(grains(indSelected).boundary,'lineWidth',4,'lineColor','gold')`
`hold off`

which I extracted from [https://mtex-toolbox.github.io/SelectingGrains.html](url)

However, when I run the first code using 'selectInteractive' function, it returns an error: `Undefined function or variable 'selectInteractive'.`

Would anybody know why this is happening? I thought the reason might be the Mtex version I am currently using (5.1.1), but I went through the Mtex Changelog and apparently this function was introduced in Mtex after the 3.3.0 version: _'-the grain selector tool for spatial grain plots was removed, nevertheless, grains still can be selected spatially.'_

**Comment by zmichels** [comment, marked as answer]
Hi there,

You’ll want to use findByLocation()

Here’s the example code from the page on that function:
plot(grains) 
p = ginput(1) 
id = findByLocation(grains,p); 
hold on, 
plot(grains(id).boundary,'linecolor','r','lineWidth',2), 
hold off
The key is to plot the grains first, then use ginput() to click on the grains and store the points. Then use those points with findByLocation to find the grains/grainId’s

Cheers,
Zach

**Reply by rtaufner** [reply]
Hi Zach, thank you for your reply! This code is running well but the problem is that I can just store one grain in the 'p' variable. As I have multiple grains around the porphyroblast, it would be interesting to select as many grains as possible. Do you have any ideia for that?

**Reply by kilir** [reply]
Zachs way for 1 or any number of grains (`p = ginput()`) is the way to go.

In addition, if you use the current version of mtex, you can click a grain (or ctrl+click multiple grains) and for now the info will be spat out in the command window. For now, you don't get the id stored but it's any easy way to find the id of a grain of interest.

![something](https://user-images.githubusercontent.com/11697557/110122053-6b753100-7dbf-11eb-8103-d4966bf0ef05.png)

**Reply by rtaufner** [reply]
Hi Kilian and Zach, Kilian's suggestion worked just fine. Thanks a lot.

Kilian, I'm having a hard time to update my Mtex to the latest versions. I spent the entire day yesterday working on the issues. I tried updating my Matlab version to the 2020b one, with no success. As I use a Mac, I also followed the instructions on [https://mtex-toolbox.github.io/installation](url), but again, the problem persisted. 

I'm almost giving up on Mac and going back to the Microsoft. 

And guys, thank you very much again for the quick response.
Cheers, 
Rhander.

**Comment by zmichels** [comment]
If you use
ginput() 
instead of ginput(1), 
you will be able to click as many points as you like until you press return on the keyboard.

**Reply by rtaufner** [reply]
Thanks Zach!!

**Reply by rtaufner** [reply]
Just one more question. The output for findByLocation is a list of grainsID 

![Screen Shot 2021-03-05 at 11 25 08](https://user-images.githubusercontent.com/75056142/110128378-791dbd00-7da5-11eb-9609-5253f2bef002.png)

Any way to save it in graind2d output so that I can proceed with orientation analyses (e.g., calculate pole figures, misorientation axes, etc)?

**Reply by kilir** [reply]
Use `grains('id',list_of_ids)`.
Cheers,
Rüdiger

**Reply by rtaufner** [reply]
Thanks Rüdiger, it worked fine!

**Reply by rtaufner** [reply]
Hi again Rüdiger and Zach, may I ask you both another question?

I am now trying to calculate low angle (0-10º) misorientation axes in the grains which I previously selected using the function `findbylocation`. The code that I am using to generate the subgrain misorientation axes is the following:

`%% Calculate misorientation axes in crystal reference frame --- https://mtex-toolbox.github.io/SubGrainBoundaries.html (subgrains)`

`close all`

`[subgrains,ebsd.grainId,ebsd.mis2mean] = calcGrains((ebsd),'angle',2*degree);`
`gB_cpx = grains.boundary('Clinopyroxene','Clinopyroxene');`

`Nmori_cpx = gB_cpx.misorientation;`

`condition = Nmori_cpx.angle/degree >= 2 & Nmori_cpx.angle/degree < 10;`

`N_axes = length(Nmori_cpxS(condition));`

`figure`
`plotAxisDistribution(Nmori_cpx(condition),'lower', 'antipodal','minmax')`
`annotate([Miller(1,0,0,ebsd('Clinopyroxene').CS,'uvw'),...`
       `   Miller(0,1,0,ebsd('Clinopyroxene').CS,'uvw'),...`
         ` Miller(0,0,1,ebsd('Clinopyroxene').CS,'uvw'),...`
 ` 'all','labeled','BackgroundColor','w','FontSize',18);`

However, I am afraid it calculates all low angle misorientation axes in the sample, and I am only interested in those misorientations from the grains previously selected. Any tips on how to overcome this problem?

**Reply by kilir** [reply]
Hi,
several options:
a) compute low angle boundaries first during computation of the grains you'll select by clicking. The subset will still contain `.innerBoundary` which are LAGB down to the threshold given by a second angle arument.
```
[grains,ebsd.grainId,ebsd.mis2mean] = calcGrains(ebsd,'angle',[10 1]*degree);
```
b) in your approach, the variable ebsd should be based on your selection
c) using the grain variable in your approach, you will always only get boundaries related to fully enclosed areas. In that case, use 
```
allBoundaries = [grains.boundary(a,a) grains(a).innerBoundary]
```
Cheers,
Rüdiger

**Reply by rtaufner** [reply]
Thanks for the help, Rüdiger!

Cheers, 
Rhander.
